File: gdal/Python/GdalShp.py
# -*- coding: utf-8 -*-
"""
@author: WZM
@time: 2021/8/22 14:22
@function:
"""
import sys

from osgeo import gdal
from osgeo import ogr


class GdalShp:
    def __init__(self):
        pass

    def getShp(self, shp_file_path):
        gdal.SetConfigOption("GDAL_FILENAME_IS_UTF8", "NO")  # 为了支持中文路径
        gdal.SetConfigOption("SHAPE_ENCODING", "")  # 为了使属性表字段支持中文

        ogr.RegisterAll()  # 注册所有驱动

        # 打开数据
        ds = ogr.Open(shp_file_path, 0)
        if ds is None:
            # print("打开%s失败" % shp_file_path)
            sys.exit(1)
        # print("文件 %s 打开成功" % shp_file_path)

        iLayerCount = ds.GetLayerCount()  # 获取该数据源中的图层个数，一般shp数据图层只有一个，mdb、dxf可能有多个
        # 获取第一个图层
        iLayerIndex = 0
        oLayer = ds.GetLayerByIndex(iLayerIndex)

        if oLayer is None:
            print("获取第 %d 个图层失败！！！" % iLayerIndex)
            sys.exit(1)

        # 对图层进行初始化
        oLayer.ResetReading()

        # 获取图层中给定属性表表头并输出
        # print("属性表结构信息")
        oDefn = oLayer.GetLayerDefn()
        iFieldCount = oDefn.GetFieldCount()
        for iAttr in range(iFieldCount):
            oField = oDefn.GetFieldDefn(iAttr)
            # print(oField.GetNameRef(), end="    ")
        # print()
        # 输出图层中的要素个数
        oFeature = oLayer.GetNextFeature()
        while oFeature is not None:
            # print("当前处理第%d个：属性值" % oFeature.GetFID())

            oGeometry = oFeature.GetGeometryRef()
            # print(oGeometry.GetArea())

            for iField in range(iFieldCount):
                oFieldDefn = oDefn.GetFieldDefn(iField)
                # print(oFeature.GetFieldAsString(iField), end="    ")
            # print()
            oFeature = oLayer.GetNextFeature()
        # print("数据集关闭！")

File: gdal/C++/gdalshp.cpp
#include "gdalshp.h"
#include "ogrsf_frmts.h"
#define  ACCEPT_USE_OF_DEPRECATED_PROJ_API_H

CGdalShp::CGdalShp()
{
}

CGdalShp::~CGdalShp()
{
}

int CGdalShp::GetShp(string shp_file_path) {
    GDALAllRegister();  //注册所有内置到GDAL/OGR中的格式驱动程序
    GDALDataset* poDS;
    CPLSetConfigOption("SHAPE_ENCODING", "");  //解决中文乱码问题
    //读取shp文件
    //const char* shp_file_path_char = shp_file_path.data();
    poDS = (GDALDataset*)GDALOpenEx(shp_file_path.c_str(), GDAL_OF_VECTOR, NULL, NULL, NULL);
    if (poDS == NULL) {
        cout << "Open failed.\n%s";
        return 0;
    }

    OGRLayer* poLayer;
    poLayer = poDS->GetLayer(0); //读取层
    OGRFeature* poFeature;

    poLayer->ResetReading();  // 开始读取该图层中的每一行
    int numFeature = 0;
    while ((poFeature = poLayer->GetNextFeature()) != NULL) {
        if (!numFeature) {
            int numField = poFeature->GetFieldCount();

            for (int inumField = 0; inumField < numField; inumField++) {
                OGRFieldDefn* poField = poFeature->GetFieldDefnRef(inumField);
                //cout << poField->GetNameRef() << "    ";
            }
            //cout << endl;
        }

        OGRGeometry* poGeometry;
        poGeometry = poFeature->GetGeometryRef();
        OGRPolygon* pPolygon = (OGRPolygon*)poGeometry->clone();
        double area = pPolygon->get_Area();
        //cout << "面积：" << pPolygon->get_Area() << endl;
        /*if (poGeometry != NULL) {
            OGRwkbGeometryType pGeoType = poGeometry->getGeometryType();
            if (pGeoType == wkbPoint) {
                OGRPoint* poPoint = (OGRPoint*)poGeometry;
                printf("%.3f,%.3f\n", poPoint->getX(), poPoint->getY());
            }else if (pGeoType == wkbPolygon){
                OGRPolygon* pPolygon = (OGRPolygon*)poGeometry->clone();
                cout << poGeometry->Polygonize() << endl;
            }
            else if (pGeoType == wkbPolygon25D) {
                OGRPolygon* pPolygon = (OGRPolygon*)poGeometry->clone();
                cout << "面积：" << pPolygon->get_Area() << endl;
            }
        }*/

        for (auto&& oField : *poFeature) {

            switch (oField.GetType()) {
            case OFTInteger:
                //cout << oField.GetInteger();
                //cout << "     ";
                break;
            case OFTInteger64:
                //cout << oField.GetInteger64();
                //cout << "     ";
                break;
            case OFTReal:
                //cout << setiosflags(ios::fixed) << oField.GetDouble();
                //cout << "     ";
                break;
            case OFTString:
                //cout << oField.GetString();
                //cout << "     ";
                break;
            default:
                //cout << oField.GetAsString();
                //cout << "     ";
                break;
            }
        }
        //cout << endl;
        numFeature++;
        OGRFeature::DestroyFeature(poFeature);

    }
    GDALClose(poDS);
}